File: console/test/ddisc/plugin/envfile_test.dart
import 'package:flutter_test/flutter_test.dart';
import 'package:retrovibed/ddisc/plugin/envfile.dart';

const _example = "FOO=\"bar\" # derp 0\n# derp 1\nBAR=\"baz\"\nBIZ=\"BAN\"\n# derp 2\n";

void main() {
  group('parseEnv', () {
    test('comment block and inline comment both attach as hints', () {
      expect(parseEnv(_example), equals([
        Variable('FOO', 'bar', 'derp 0'),
        Variable('BAR', 'baz', 'derp 1'),
        Variable('BIZ', 'BAN', ''),
      ]));
    });

    test('empty content yields no variables', () {
      expect(parseEnv(''), isEmpty);
    });

    test('unquoted value with no comment', () {
      expect(parseEnv('FOO=bar\n'), equals([Variable('FOO', 'bar', '')]));
    });

    test('blank line clears pending comment', () {
      expect(parseEnv('# orphaned\n\nFOO=bar\n'), equals([Variable('FOO', 'bar', '')]));
    });
  });

  group('serializeEnv', () {
    test('empty list yields empty content', () {
      expect(serializeEnv([]), equals(''));
    });

    test('serializes only the given variables, dropping anything else', () {
      final got = serializeEnv([Variable('FOO', 'bar', ''), Variable('B', '', '')]);
      expect(got, equals('FOO=bar\nB=\n'));
    });

    test('quotes values and preserves hints as inline comments', () {
      final got = serializeEnv([Variable('FOO', 'has space', 'derp')]);
      expect(got, equals('FOO="has space" # derp\n'));
    });

    test('round trip: parseEnv(serializeEnv(variables)) reflects exactly the given variables', () {
      final variables = [Variable('FOO', 'bar', 'derp 0'), Variable('BIZ', 'BAN', '')];
      expect(parseEnv(serializeEnv(variables)), equals(variables));
    });
  });
}

File: retroapi/envfile/envfile_test.go
package envfile_test

import (
	"testing"

	"github.com/retrovibed/retrovibed/retroapi/envfile"
	"github.com/stretchr/testify/require"
)

const example = "FOO=\"bar\" # derp 0\n# derp 1\nBAR=\"baz\"\nBIZ=\"BAN\"\n# derp 2\n"

func TestParse(t *testing.T) {
	t.Run("comment block and inline comment both attach as hints", func(t *testing.T) {
		require.Equal(t, []envfile.Variable{
			{Key: "FOO", Value: "bar", Hint: "derp 0"},
			{Key: "BAR", Value: "baz", Hint: "derp 1"},
			{Key: "BIZ", Value: "BAN", Hint: ""},
		}, envfile.Parse(example))
	})

	t.Run("empty content yields no variables", func(t *testing.T) {
		require.Empty(t, envfile.Parse(""))
	})

	t.Run("unquoted value with no comment", func(t *testing.T) {
		require.Equal(t, []envfile.Variable{
			{Key: "FOO", Value: "bar", Hint: ""},
		}, envfile.Parse("FOO=bar\n"))
	})

	t.Run("blank line clears pending comment", func(t *testing.T) {
		require.Equal(t, []envfile.Variable{
			{Key: "FOO", Value: "bar", Hint: ""},
		}, envfile.Parse("# orphaned\n\nFOO=bar\n"))
	})
}

func TestApply(t *testing.T) {
	t.Run("edits an existing value, preserving its comment", func(t *testing.T) {
		got := envfile.Apply(example, []envfile.Variable{
			{Key: "FOO", Value: "updated"},
		})
		require.Equal(t, "FOO=updated # derp 0\n# derp 1\nBAR=\"baz\"\nBIZ=\"BAN\"\n# derp 2\n", got)
	})

	t.Run("appends an edit whose key has no matching line", func(t *testing.T) {
		got := envfile.Apply("FOO=bar\n", []envfile.Variable{
			{Key: "NEW", Value: "value"},
		})
		require.Equal(t, "FOO=bar\nNEW=value\n", got)
	})

	t.Run("quotes values containing whitespace or a comment character", func(t *testing.T) {
		got := envfile.Apply("FOO=bar\n", []envfile.Variable{
			{Key: "FOO", Value: "has space"},
		})
		require.Equal(t, "FOO=\"has space\"\n", got)
	})

	t.Run("round trip: parse(apply(content, edits)) reflects the edits", func(t *testing.T) {
		edits := []envfile.Variable{{Key: "FOO", Value: "changed"}}
		got := envfile.Parse(envfile.Apply(example, edits))
		require.Equal(t, []envfile.Variable{
			{Key: "FOO", Value: "changed", Hint: "derp 0"},
			{Key: "BAR", Value: "baz", Hint: "derp 1"},
			{Key: "BIZ", Value: "BAN", Hint: ""},
		}, got)
	})
}
